### ISSUE TITLE
Overloaded constructors having same class name in arguments are buggy

### ISSUE THREAD

**Issue by kysmith-csg** [eclipse-jdt/eclipse.jdt.ui#1165]
I saw this while working with [eclipse/gef-classic.](https://github.com/eclipse/gef-classic/).

The class `org.eclipse.draw2d.geometry.Rectangle` has two constructors of note, written as follows (comment added to the first one for clarity):

```java
public Rectangle(/*org.eclipse.draw2d.geometry.*/Rectangle rect) {
```

and
```java
public Rectangle(org.eclipse.swt.graphics.Rectangle rect) {
```

So far I have seen this lead to 3 buggy scenarios:
1. In code that calls the 2nd constructor (the SWT Rectangle), if you **Open Declaration** it will open the first one (draw2d Rectangle).

https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/136501322/ef096da2-dd4d-45c8-9750-5539ed99af42

2. The **Outline View** in the Rectangle class does not let you navigate to the SWT Rectangle constructor (note that it doesn't highlight it, if it wasn't visible already it wouldn't be scrolled to).

https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/136501322/29a2bf4d-e0ee-4137-9aff-7a6cc464e0e0

3. **Quick Outline** in the Rectangle class does not let you navigate to the SWT rectangle constructor (probably same bug as above)

Basically in all cases where you would want to see the SWT Rectangle constructor, you are always given the other one instead.

---

	Eclipse.org	Java Development Tools UI	3.31.0.v20231115-2128	org.eclipse.jdt.ui

**Comment by jukzi** [comment]
please provide a self containing minimal reproducer

**Comment by kysmith-csg** [comment]
[jdt-ui.issue.1165.zip](https://github.com/eclipse-jdt/eclipse.jdt.ui/files/14166664/jdt-ui.issue.1165.zip)

Projects attached. The main project is `jdt-ui.issue.1165.main`, the other ones are just the libraries. 
Steps:
1. Import `jdt-ui.issue.1165.main` into Eclipse
2. Add a classpath entry for `jdt-ui.issue.1165.1.jar` and `jdt-ui.issue.1165.2.jar`.
3. Add the corresponding source attachments.
4. In `Main.java` of your project, `Open Declaration` of the `MyClass` constructors and see that the classfile editor doesn't select the correct one (the outline view does).

So I guess this reproducer helps me to revise the issue. The issue seems to be with the `class file editor`. Having the other projects imported in the workspace cause the functionalities to behave without bugs as expected. It seems to be only when they are class files that the source editor has trouble selecting the correct one.

**Comment by jukzi** [comment]
can you provide a junit test?

**Comment by kysmith-csg** [comment]
@jukzi  I don't know how to do UI testing, and I don't really have much time at the moment due to my work. I can try to do it at some point but I have no idea when that would be.
